src/combinators/mod.rs

```diff
@@ -974,7 +974,10 @@ impl BackToTheBuffer for &mut [u8] {
     fn reserve_write_use<Tmp, Gen: Fn(Self) -> GenResult<(Self, Tmp)>, Before: Fn(Self, Tmp) -> GenResult<Self>>(self, reserved: usize, gen: Gen, before: Before) -> GenResult<Self> {
         let (res, buf) = self.split_at_mut(reserved);
         let (buf, tmp) = gen(buf)?;
-        before(res, tmp)?;
+        let res = before(res, tmp)?;
+        if !res.is_empty() {
+            return Err(GenError::BufferTooBig(res.len()))
```

Comment:
I'd personally simply panic here. It's not an error that can be handled by the caller in any meaningful way, it's simply the `before` serializer not fullfilling the API requirements here.

Reply:
I *hate* libraries panicing, should be the caller's choice imo.
The caller can chose to panic when getting this error, or can decide that "it's fine, I already did a zero-pass before anyways", for example

Reply:
I disagree in this case. The caller might only call some very high level serializer and will have no idea if this error being returned can be ignored simply because the whole data was initialized with zeroes before.

Also in this case, not panicking here makes it impossible to safely implement the trait for `Vec<u8>` as you would leak uninitialized memory into safe code.

Reply:
@Geal What do you think?

Note that if we want, we can still call panic! for the specific case of Vec

Reply:
> Note that if we want, we can still call panic! for the specific case of Vec

Indeed but it would seem confusing if the same serializer behaves that different depending on the writer that is passed in.

Did you try implementing it for `Vec` btw? I'm curious if that actually fits with the current trait and what kind of ugly unsafe code we need to make it work :)

Reply:
Two unsafe calls are needed, couldn't find a way with less than that

Reply:
Now there's only one.
We do have to copy over some data once more, but it's less prone to insane errors

Reply:
I'll leave it as is for now, maybe you can open a separate issue if you really think we should panic there